### PR TITLE
Validate members type attribute

### PR DESCRIPTION
**Pull request by nipunthilakshan**
## Purpose
Fix https://github.com/wso2/product-is/issues/10972

### CODE REVIEW

File: modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java [hunk 1]
```diff
@@ -113,6 +116,26 @@ private static void validateSCIMObjectForRequiredSubAttributes(AbstractAttribute
                             }
                         }
                     }
+
+                    // Check for canonical attributes in groups.
+                    List<String> canonicalValues = subAttributeSchema.getCanonicalValues();
+                    if (scimObject instanceof Group && canonicalValues != null) {
```
Comment (somindatommy) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java:122]: what happens if the list is empty? can we validate the empty list scenario?

File: modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java [hunk 2]
```diff
@@ -113,6 +116,26 @@ private static void validateSCIMObjectForRequiredSubAttributes(AbstractAttribute
                             }
                         }
                     }
+
+                    // Check for canonical attributes in groups.
+                    List<String> canonicalValues = subAttributeSchema.getCanonicalValues();
```
Comment (somindatommy) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java:121]: can we extract this for a private method? WDYT?

**Comment by somindatommy** [comment]
@nipunthilakshan after merging this PR u need to bump the Charon version in the IS as well

File: modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java (hunk 1 above)
Comment (nipunthilakshan) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java]: addressed

File: modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java (hunk 2 above)
Comment (nipunthilakshan) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java:122]: addressed, please check and confirm

File: modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java [hunk 3]
```diff
@@ -113,6 +117,14 @@ private static void validateSCIMObjectForRequiredSubAttributes(AbstractAttribute
...
+                    if (!canonicalValues.isEmpty()) {
```
Comment (somindatommy) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java:123]: If the `canonicalValues` returned null, won't this throw a null pointer exception? Can we use `CollectionUtils.isNotEmpty`?

Comment (somindatommy) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java:123]: We can also move this to the method right?

File: modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java [hunk 4]
```diff
@@ -128,7 +140,40 @@ private static void validateSCIMObjectForRequiredSubAttributes(AbstractAttribute
...
+     * @param canonicalValues
+     * @throws CharonException
+     * @throws BadRequestException
+     */
+    private static void validateSCIMObjectForCanonicalAttributes(AbstractAttribute attribute,
+                                                                 AttributeSchema subAttributeSchema,
+                                                                 AbstractSCIMObject scimObject,
+                                                                 List<String> canonicalValues) throws
+            CharonException, BadRequestException {
+        if (scimObject instanceof Group && canonicalValues != null) {
```
Comment (somindatommy) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java:165]: Shall we take the negation of this else block and end the method from there?

File: modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java [hunk 5]
```diff
@@ -128,7 +140,40 @@ private static void validateSCIMObjectForRequiredSubAttributes(AbstractAttribute
...
+     * Validate SCIMObject for canonical attributes in groups.
+     *
+     * @param attribute
+     * @param subAttributeSchema
+     * @param scimObject
+     * @param canonicalValues
+     * @throws CharonException
+     * @throws BadRequestException
+     */
+    private static void validateSCIMObjectForCanonicalAttributes(AbstractAttribute attribute,
```
Comment (somindatommy) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java:160]: `validateConicalAttributesInScimObject` WDYT?

File: modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java [hunk 6]
```diff
@@ -128,7 +140,40 @@ private static void validateSCIMObjectForRequiredSubAttributes(AbstractAttribute
...
+
+    /*
+     * Validate SCIMObject for canonical attributes in groups.
+     *
+     * @param attribute
+     * @param subAttributeSchema
+     * @param scimObject
+     * @param canonicalValues
+     * @throws CharonException
+     * @throws BadRequestException
```
Comment (somindatommy) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AbstractValidator.java:158]: If we are adding a method comment, make sure to complete it with appropriate descriptions.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AttributeSchema.java [hunk 7]
```diff
@@ -71,5 +72,9 @@
 
     public void removeSubAttribute(String subAttributeName) throws CharonException;
 
+    public List<String> getCanonicalValues();
+
+    public void setCanonicalValues(ArrayList<String> canonicalValues);
+
```
Comment (somindatommy) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AttributeSchema.java:78]: Don't the methods in an interface always public? Do we need to declare them as public?

File: modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AttributeSchema.java [hunk 8]
```diff
@@ -71,5 +72,15 @@
 
     public void removeSubAttribute(String subAttributeName) throws CharonException;
 
+    /**
+     * @return Get list of canonical values.
+     */
+    List<String> getCanonicalValues();
+
+    /**
+     * @param canonicalValues canonical values to set.
```
Comment (somindatommy) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AttributeSchema.java:81]: Method descriptions are missing? Shall we add it?

Comment (nipunthilakshan) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/schema/AttributeSchema.java:83]: addressed

**Review by AnuradhaSK — APPROVED** [review]

**Review by somindatommy — APPROVED** [review]

**Pull request merged**